vowpalwabbit/cats.cc

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "cats.h"
+#include "parse_args.h"
+#include "err_constants.h"
+#include "api_status.h"
+#include "debug_log.h"
+
+// Aliases
+using VW::LEARNER::single_learner;
+using std::endl;
+using VW::cb_continuous::continuous_label;
+using VW::cb_continuous::continuous_label_elm;
+using VW::config::make_option;
+using VW::config::option_group_definition;
+using VW::config::options_i;
+
+// Enable/Disable indented debug statements
+VW_DEBUG_ENABLE(false)
+
+// Forward declarations
+namespace VW
+{
+  void finish_example(vw& all, example& ec);
+}
+
+namespace VW { namespace continuous_action { namespace cats {
+  ////////////////////////////////////////////////////
+  // BEGIN cats reduction and reduction methods
+  struct cats
+  {
+    cats(single_learner* p_base);
+
+    int learn(example& ec, api_status* status);
+    int predict(example& ec, api_status* status);
+
+  private:
+    single_learner* _base = nullptr;
+  };
+
+  // Pass through
+  int cats::predict(example& ec, api_status*)
+  {
+    VW_DBG(ec) << "cats::predict(), " << features_to_string(ec) << endl;
+    _base->predict(ec);
```

Comment:
I considered try/catch handler here and discarded the idea in favor of letting the exception get handled upstream.

My thinking is the following:
A) Current behavior
A.1) If base.predict() does not throw, there is no issue.
A.2) If base.predict() throws then somewhere upstream vw is handling this already.  If this behavior is changed to  to return an error code, current error handling logic breaks.  So it's better to let vw handle it upstream.  

B) Future behavior - base->predict() will not throw and only return error codes and this can be propagated up the stack.  This change will happen after we start supporting error codes properly.